webauthn-server-core/src/main/java/com/yubico/webauthn/data/Extensions.java

```diff
@@ -108,6 +114,296 @@ public Optional<Boolean> getRk() {
+     * RegistrationExtensionInputs.RegistrationExtensionInputsBuilder#credProtect(CredentialProtectionInput)
+     * credProtect} extension is not set in the request, this has no effect.
+     *
+     * <p>If the {@link
+     * RegistrationExtensionInputs.RegistrationExtensionInputsBuilder#credProtect(CredentialProtectionInput)
+     * credProtect} extension is set in the request with {@link
+     * CredentialProtectionInput#isEnforceCredentialProtectionPolicy()
+     * enforceCredentialProtectionPolicy} set to <code>false</code>, this has no effect.
+     *
+     * <p>If the {@link
+     * RegistrationExtensionInputs.RegistrationExtensionInputsBuilder#credProtect(CredentialProtectionInput)
+     * credProtect} extension is set in the request with {@link
+     * CredentialProtectionInput#isEnforceCredentialProtectionPolicy()
+     * enforceCredentialProtectionPolicy} set to <code>true</code>, then this throws an {@link
+     * IllegalArgumentException} if the <code>credProtect</code> authenticator extension output does
+     * not equal the {@link CredentialProtectionInput#getCredentialProtectionPolicy()
+     * credentialProtectionPolicy} set in the request.
+     *
+     * <p>This function is called automatically in {@link
+     * RelyingParty#finishRegistration(FinishRegistrationOptions)} when the request has {@link
+     * CredentialProtectionInput#isEnforceCredentialProtectionPolicy()
+     * enforceCredentialProtectionPolicy} is set to <code>true</code>; you should not need to call
+     * it yourself.
+     *
+     * @param request the arguments to start the registration ceremony.
+     * @param response the response from the registration ceremony.
+     * @throws IllegalArgumentException if the {@link
+     *     RegistrationExtensionInputs.RegistrationExtensionInputsBuilder#credProtect(CredentialProtectionInput)
+     *     credProtect} extension is set in the request with {@link
+     *     CredentialProtectionInput#isEnforceCredentialProtectionPolicy()
+     *     enforceCredentialProtectionPolicy} set to <code>true</code> and the <code>credProtect
+     *     </code> authenticator extension output does not equal the {@link
+     *     CredentialProtectionInput#getCredentialProtectionPolicy() credentialProtectionPolicy} set
+     *     in the request.
+     * @see <a
+     *     href="https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#sctn-credProtect-extension">CTAP2
+     *     §12.1. Credential Protection (credProtect)</a>
+     */
+    public static void validateExtensionOutput(
+        PublicKeyCredentialCreationOptions request,
+        PublicKeyCredential<AuthenticatorAttestationResponse, ClientRegistrationExtensionOutputs>
+            response) {
+      request
+          .getExtensions()
+          .getCredProtect()
+          .ifPresent(
+              credProtect -> {
+                if (credProtect.isEnforceCredentialProtectionPolicy()) {
```

Comment:
From out-of-band discussion: since this is inspired by the [client processing behaviour](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#sctn-credProtect-extension), the logic here should match the client logic. Therefore this should not apply when the chosen policy is `userVerificationOptional`:
> When `enforceCredentialProtectionPolicy` is true, and `credentialProtectionPolicy`'s value is either _userVerificationOptionalWithCredentialIDList_ or _userVerificationRequired_, the platform SHOULD NOT create the credential in a way that does not implement the requested protection policy.